probe-rs/src/session.rs

```diff
@@ -1,14 +1,121 @@
 use crate::config::target::Target;
-use crate::probe::MasterProbe;
+use crate::probe::{DebugProbeError, MasterProbe};
 
 pub struct Session {
     pub target: Target,
     pub probe: MasterProbe,
+
+    hw_breakpoint_enabled: bool,
+    active_breakpoints: Vec<Breakpoint>,
+    bp_id_count: usize,
 }
 
 impl Session {
     /// Open a new session with a given debug target
     pub fn new(target: Target, probe: MasterProbe) -> Self {
-        Self { target, probe }
+        Self {
+            target,
+            probe,
+            hw_breakpoint_enabled: false,
+            active_breakpoints: Vec::new(),
+            bp_id_count: 0,
+        }
+    }
+
+    /// Set a hardware breakpoint
+    pub fn set_hw_breakpoint(&mut self, address: u32) -> Result<BreakpointId, DebugProbeError> {
+        log::debug!("Trying to set HW breakpoint at address {:#08x}", address);
+
+        // Get the number of HW breakpoints available
+        let num_hw_breakpoints =
+            self.target
+                .core
+                .get_available_breakpoint_units(&mut self.probe)? as usize;
+
+        log::debug!("{} HW breakpoints are supported.", num_hw_breakpoints);
+
+        if num_hw_breakpoints <= self.active_breakpoints.len() {
+            // We cannot set additional breakpoints
+            log::warn!("Maximum number of breakpoints ({}) reached, unable to set additional HW breakpoint.", num_hw_breakpoints);
+
+            // TODO: Better error here
+            return Err(DebugProbeError::UnknownError);
+        }
+
+        if !self.hw_breakpoint_enabled {
+            self.target.core.enable_breakpoints(&mut self.probe, true)?;
+            self.hw_breakpoint_enabled = true;
+        }
+
+        let bp_unit = self.find_free_breakpoint_unit();
+
+        log::debug!("Using comparator {} of breakpoint unit", bp_unit);
+        // actually set the breakpoint
+        self.target
+            .core
+            .set_breakpoint(&mut self.probe, bp_unit, address)?;
+
+        let id = BreakpointId(self.bp_id_count);
+        self.bp_id_count += 1;
```

Comment:
Any reason why you don't just use the address as a key/id in a HashMap? Because that way you could clear the breakpoint rightaway. With the current implementation you will need a HashMap<Address, Id> as the user of this API, which does seem cumbersome.

Reply:
I guess it doesn't really matter if I use a separate id or the address itself, but for me a separate id seems a bit cleaner, and a bit more robust against changes in the future.

I think for the very limited amount of HW breakpoints available, the storage doesn't really matter. Would your suggestion be to not return any kind of id, and just use hardware addresses to track break points?

Reply:
> Would your suggestion be to not return any kind of id, and just use hardware addresses to track break points?

Basically this, yes. I don't mind the storage type, I just assumed this concept goes well with a HashMap. For me the end user just wants to set/clear a BP at address 0xxxxxxxxx and not deal with any intrinsics :)

Reply:
I was more thinking about future enhancements, but I think for now using addresses is good enough.